src/contrib/cluster/Akka.Cluster.Metrics/Routing/MetricSelectors.cs

```diff
@@ -0,0 +1,239 @@
+                default:
+                    return DynamicAccess.CreateInstanceFor<IMetricsSelector>(selectorTypeName, config)
+                        .Recover(ex => throw new ArgumentException($"Cannot instantiate metrics-selector [{selectorTypeName}]," +
+                                                                   $"make sure it extends [Akka.Cluster.Metrics.MetricsSelector] and " +
+                                                                   $"has constructor with [Akka.Configuration.Config] parameter", ex))
+                        .Get();
+            }
+        }
+    }
+
+    /// <summary>
+    /// A MetricsSelector producing weights from remaining capacity.
+    /// The weights are typically proportional to the remaining capacity.
+    /// </summary>
+    public abstract class CapacityMetricsSelector : IMetricsSelector
+    {
+        /// <summary>
+        /// Remaining capacity for each node. The value is between 0.0 and 1.0, where 0.0 means no remaining capacity
+        /// (full utilization) and 1.0 means full remaining capacity (zero utilization).
+        /// </summary>
+        public abstract IImmutableDictionary<Actor.Address, decimal> Capacity(IImmutableSet<NodeMetrics> nodeMetrics);
+
+        /// <summary>
+        /// Converts the capacity values to weights. The node with lowest capacity gets weight 1
+        /// (lowest usable capacity is 1%) and other nodes gets weights proportional to their capacity compared to
+        /// the node with lowest capacity.
+        /// </summary>
+        public IImmutableDictionary<Actor.Address, int> Weights(IImmutableDictionary<Actor.Address, decimal> capacity)
+        {
+            if (capacity.Count == 0)
+                return ImmutableDictionary<Actor.Address, int>.Empty;
+
+            var min = capacity.Min(c => c.Value);
+            // lowest usable capacity is 1% (>= 0.5% will be rounded to weight 1), also avoids div by zero
+            var divisor = Math.Max(0.01M, min);
+            return capacity.ToImmutableDictionary(pair => pair.Key, pair => (int)Math.Round(pair.Value / divisor));
+        }
+        
+        /// <inheritdoc />
+        public IImmutableDictionary<Actor.Address, int> Weights(IImmutableSet<NodeMetrics> nodeMetrics)
+        {
+            return Weights(Capacity(nodeMetrics));
+        }
+    }
+
+    /// <summary>
+    /// MetricsSelector that uses the heap metrics.
+    /// Low heap capacity => small weight.
+    /// </summary>
+    public class HeapMetricsSelector : CapacityMetricsSelector
+    {
+        /// <summary>
+        /// Singleton instance
+        /// </summary>
+        public static readonly HeapMetricsSelector Instance = new HeapMetricsSelector();
+        
+        /// <inheritdoc />
+        public override IImmutableDictionary<Actor.Address, decimal> Capacity(IImmutableSet<NodeMetrics> nodeMetrics)
+        {
+            return nodeMetrics
+                .Select(StandardMetrics.HeapMemory.Unapply)
```

Comment:
There is only `Unapply`, which is basically converting data object to tuple containing some of it's fields. Or to the `Option<ValueTuple>` to be more accurate, since requested metrics may be missing.
Have just renamed it to `Decompose` to reflect the semantics.